tests/mail/tests.py

```diff
@@ -1291,6 +1326,92 @@ def test_sanitize_address_header_injection(self):
+            ("<to@xn--10cl1a0b660p.example.com>", "<to@xn--10cl1a0b660p.example.com>"),
+            (
+                '"Display, Name" <to@xn--nxasmm1c.example.com>',
+                '"Display, Name" <to@xn--nxasmm1c.example.com>',
+            ),
+            # Non-ASCII display-name as RFC-2047 encoded-word.
+            (
+                "Tó Example <to@example.com>",
+                "=?utf-8?q?T=C3=B3_Example?= <to@example.com>",
+            ),
+            # Addresses with two @ signs (quoted-string localpart).
+            ('"to@other.com"@example.com', '"to@other.com"@example.com'),
+            (
+                'To Example <"to@other.com"@example.com>',
+                'To Example <"to@other.com"@example.com>',
+            ),
+            # Addresses with long non-ASCII display names.
+            (
+                "Tó Example very long" * 4 + " <to@example.com>",
+                "=?utf-8?q?T=C3=B3_Example_very_longT=C3=B3_Example_very_longT"
+                "=C3=B3_Example_?="
+                " =?utf-8?q?very_longT=C3=B3_Example_very_long?= <to@example.com>",
+            ),
+            # Address with long display name and non-ASCII domain.
+            (
+                "To Example very long" * 4 + " <to@exampl€.com>",
+                "To Example very longTo Example very longTo Example very lo"
+                "ngTo Example very long <to@xn--exampl-nc1c.com>",
+            ),
+        ]
+        for address, expected in cases:
+            with self.subTest(address=address):
+                email = EmailMessage(to=[address])
+                actual = email.message().as_bytes().decode()
+                # Unfold FWS and extract the To header. (This is not even close
+                # to a complete header parser, but is sufficient for this test.
+                # Note it does not recombine adjacent/folded RFC 2047 encoded-words.)
+                headers = re.sub(r"\s*\r?\n ", " ", actual).splitlines()
+                to_header = [h for h in headers if h.startswith("To:")][0]
+                expected_header = f"To: {expected}"
+                self.assertEqual(to_header, expected_header)
+
+    def test_address_header_injection(self):
+        # TODO: modern-email: expected error msg from Python email will be:
+        #   "Header values may not contain linefeed or carriage return characters"
+        #   (Current error comes from forbid_multi_line_headers().)
+        msg = "Header values can't contain newlines"
+        cases = [
+            "Name\nInjection <to@example.com>",
+            '"Name\nInjection" <to@example.com>',
+            '"Name\rInjection" <to@example.com>',
+            '"Name\r\nInjection" <to@example.com>',
+            "Name <to\ninjection@example.com>",
+            "to\ninjection@example.com",
+        ]
+        for email_address in cases:
+            with self.subTest(email_address=email_address):
+                email = EmailMessage(to=[email_address])
```

Comment:
Should we test other headers here too, in addition to "To:"?

Reply:
This test will replace test_sanitize_address_header_injection() above, which unit tests the (to be removed) sanitize_address_header() internal function. Address header sanitization is built into Python's modern email API, so will no longer be directly enforced in Django email.

I was trying to balance "preserving the spirit" of the existing Django mail tests vs. re-testing Python's own code (which has extensive tests in Python itself).

I'm not opposed to adding a subTest loop over all the other address headers, but I'm also not sure it adds much value over Python's own test cases. (Or we could just skip this test entirely, leaving it to Python.)

Reply:
OK, then I'd probably leave this test in as it is (at least for now).

Reply:
Actually, the more I think about this, I agree it's a good idea to be thorough. I'll add a nested subtest over all the structured address headers.